### PR TITLE
feat(all): implement android library and tweak ios and react

### PR DESCRIPTION
**Pull request by liambutler-lawrence**

### CODE REVIEW

File: packages/npm/send/android/build.gradle [hunk 1]
```diff
@@ -90,5 +95,8 @@ dependencies {
   //noinspection GradleDynamicVersion
   implementation "com.facebook.react:react-native:+"
   implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
-}
+  implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.6.0"
+  implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.8.1-Beta"
+  implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.8.1-Beta"
```
Comment (nikwotton) [inline packages/npm/send/android/build.gradle:100]: Whilst I'm generally a fan of beta versions of things I'm going to ask anyways - why are you using beta here?

File: packages/npm/send/src/index.tsx [hunk 2]
```diff
@@ -32,7 +32,7 @@ export type Request = {
     | 'OPTIONS'
     | 'CONNECT'
     | 'TRACE';
-  body?: string;
+  body: string | null;
```
Comment (nikwotton) [inline packages/npm/send/src/index.tsx:35]: What's the difference between these two?

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 3]
```diff
@@ -1,25 +1,173 @@
...
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.internal.http.HttpMethod
+import okio.Buffer
+import okio.ByteString.Companion.decodeBase64
+import okio.GzipSource
+import okio.IOException
+import java.util.concurrent.TimeUnit
+
+// FIXME: Actually check charset parameter first if set, then fall back to MIME type
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:22]: no ticket number?

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 4]
```diff
@@ -1,25 +1,173 @@
...
+private fun bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: List<String>): Boolean {
+  return contentTypeHeader != null && utf8ContentTypes.contains(contentTypeHeader.split(';')[0])
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:24]: Not sure if this is actually better, but at least it's shorter.....up to you which to use, just making the possibility known
```suggestion
  return utf8ContentTypes.contains(contentTypeHeader?.split(';')?.get(0))
```

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 5]
```diff
@@ -1,25 +1,173 @@
...
+}
+
+@Serializable
+data class Request(
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:28]: You're gonna end up with confusion if you use this generic/common of a name, just a heads up...

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 6]
```diff
@@ -1,25 +1,173 @@
...
+  val baseURL: String,
+  val path: String,
+  val queryParameters: Map<String, String>,
+  val headerParameters: Map<String, String>,
+  val method: String,
+  val body: String?,
+  val utf8ContentTypes: List<String>,
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:35]: You could probably stand to add some defaults to these
```suggestion
  val path: String = "",
  val queryParameters: Map<String, String> = emptyMap(),
  val headerParameters: Map<String, String> = emptyMap(),
  val method: String,
  val body: String? = null,
  val utf8ContentTypes: List<String> = emptyList(),
```

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 7]
```diff
@@ -1,25 +1,173 @@
...
+) {
+  val httpUrl: HttpUrl
+    get() {
+      val urlBuilder = (baseURL + path).toHttpUrl().newBuilder()
+      queryParameters.forEach { (key, value) ->
+        urlBuilder.addQueryParameter(key, value)
+      }
+
+      return urlBuilder.build()
+    }
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:45]: Doing this will make httpUrl be re-calculated every time you call it. Since the inputs are all vals, seems like you could just calculate it once the first time you reference it then cache the result using lazy
```suggestion
  val httpUrl: HttpUrl
    by lazy {
      val urlBuilder = (baseURL + path).toHttpUrl().newBuilder()
      queryParameters.forEach { (key, value) ->
        urlBuilder.addQueryParameter(key, value)
      }

      urlBuilder.build()
    }
```

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 8]
```diff
@@ -1,25 +1,173 @@
...
+    get() {
+      val requestBuilder = okhttp3.Request.Builder().url(httpUrl)
+      headerParameters.forEach { (key, value) ->
+        requestBuilder.addHeader(key, value)
+      }
+
+      requestBuilder.method(
+        method,
+        if (body == null) if (HttpMethod.requiresRequestBody(method)) "".toRequestBody() else null
+        else {
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:57]: As is, this took me a minute to parse..
```suggestion
        if (body == null) 
            if (HttpMethod.requiresRequestBody(method)) "".toRequestBody() else null
        else {
```

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 9]
```diff
@@ -1,25 +1,173 @@
...
+
+      return requestBuilder.build()
+    }
+}
+
+@Serializable
+data class Response(
+  val statusCode: Int,
+  val headerParameters: Map<String, String>,
+  val body: String?,
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:79]: Same comments as above

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 10]
```diff
@@ -1,25 +1,173 @@
...
+        gzipSource.close()
+        if (bodyIsUTF8) buffer.readUtf8() else buffer.readByteString().base64()
+      } else {
+        if (bodyIsUTF8) responseBody.string() else responseBody.byteString().base64()
+      }
+    }
+  )
+}
+
+class SendModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:108]: Classes should generally have their own files btw

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 11]
```diff
@@ -1,25 +1,173 @@
...
+
+  private val client =
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:110]: If you use by lazy here, you can move initialization to when this is needed (assuming you want that)

File: packages/npm/send/android/build.gradle (hunk 1 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/build.gradle:100]: Copied it from the installation docs on https://github.com/Kotlin/kotlinx.coroutines#gradle. Seemed strange to me too.

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 12]
```diff
@@ -1,25 +1,173 @@
...
     return NAME
   }
 
-  // Example method
-  // See https://reactnative.dev/docs/native-modules-android
   @ReactMethod
-  fun multiply(a: Double, b: Double, promise: Promise) {
-    promise.resolve(a * b)
+  fun send(stringifiedRequest: String, promise: Promise) {
+//    CoroutineScope(Dispatchers.IO).launch {
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:125]: This comment staying intentionally?

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt [hunk 13]
```diff
@@ -1,25 +1,173 @@
...
+          serializationException
+        )
+        return
+      }
+
+      val okHttpResponse: okhttp3.Response
+      try {
+        okHttpResponse = client.newCall(request.okHttpRequest).execute()
+      } catch (ioException: IOException) {
+        // FIXME: switch on error and reject promise
```
Comment (nikwotton) [inline packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:143]: no ticket number?

File: packages/npm/send/ios/Send.swift [hunk 14]
```diff
@@ -1,17 +1,41 @@
+private let requestError = "Your request was invalid. Please make sure it conforms to the Request type and try again."
+private let serverError = "Your request did not receive a response. Please make sure you are connected to the Internet and try again."
+private let responseError = "Your request received a response, but it couldn't be processed. Please file an issue on GitHub or try again."
+private let unknownError = "Something went wrong. Please file an issue on GitHub or try again."
+
 private func bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: [String]) -> Bool {
-    guard let contentType = contentTypeHeader?.split(separator: ";").first else {
+    guard let contentType = contentTypeHeader?.split(separator: ";")[0] else {
```
Comment (nikwotton) [inline packages/npm/send/ios/Send.swift:7]: just a curiosity - why this change?

File: packages/npm/send/android/build.gradle (hunk 1 above)
Comment (nikwotton) [reply packages/npm/send/android/build.gradle:100]: well that is just a suggestion - looks like 1.8.0 is stable if you want to use it. Note I do generally trust betas from jetbrains so no worries if you leave it, just calling it out in case you hadn't noticed it

File: packages/npm/send/src/index.tsx (hunk 2 above)
Comment (liambutler-lawrence) [reply packages/npm/send/src/index.tsx:35]: It's the difference between a) a property that may or may not exist in the (serialized) object at all, vs b) a property that exists regardless, but whose value may not exist. It's still confusing because the terminology & syntax for these concepts is different in all 3 languages, and oftentimes conflicting, but I actually just made a table while working on this PR.

| Language | Key may exist or not | Value may exist or not |
|:---|:---|:---|
| TypeScript | `body: string \| null` | `body?: string` |
| Kotlin | `body: String?` | `body: String? = null` |
| Swift | **DOES NOT EXIST** | `body: String?` |

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 3 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:22]: Not for our OSS projects.

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 7 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:45]: Cool, thanks!

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 9 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:79]: See response above.

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 13 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:143]: See response above.

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 12 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:125]: It's an implicit FIXME for the next PR. I'm currently using OKHTTP's `.execute()`, which is documented to block the thread and works in the example app, but would presumably cause lots of problems in prod.

File: packages/npm/send/src/index.tsx (hunk 2 above)
Comment (nikwotton) [reply packages/npm/send/src/index.tsx:35]: fun stuff

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 3 above)
Comment (nikwotton) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:22]: boooo :P

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 12 above)
Comment (nikwotton) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:125]: I'm generally against FIXMEs being implicit, though since you're the only dev here I'll allow it.... ;)

File: packages/npm/send/ios/Send.swift (hunk 14 above)
Comment (gtokman) [reply packages/npm/send/ios/Send.swift:7]: Same q!

File: packages/npm/send/android/build.gradle (hunk 1 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/build.gradle:100]: Got it, will switch. NPM packages usually put the latest stable version in the installation instructions 😅

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 5 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:28]: My understanding is Java/Kotlin classes are namespaced automatically, is that not the case? This library isn't meant to be consumed directly from other native code, but rather, only from React Native.

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 6 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:35]: Going to save this for a v2. Feel free to file an issue 😉

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 10 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:108]: Does that include data classes, or only proper classes? How about helper functions (like at the top of the file)?

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 11 above)
Comment (liambutler-lawrence) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:110]: Thanks! Updated.

File: packages/npm/send/ios/Send.swift (hunk 14 above)
Comment (liambutler-lawrence) [reply packages/npm/send/ios/Send.swift:7]: Good catch, reverted!

**Review by gtokman — APPROVED** [review]

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 5 above)
Comment (nikwotton) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:28]: Runtime you're fine and the code will have no issue with it - when I said confusion I meant developer confusion ;)

File: packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt (hunk 10 above)
Comment (nikwotton) [reply packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt:109]: Generally yes, though when you have nested data classes each of which only has one or two fields it's probably fine to put them in the same file. I'd make a separate file for the data classes from the other classes though. 
As for helper functions, it depends on how many there are but often times if they don't fit within a class I'll make a Utils.kt file
